Issue-76: Improve deserialization error messages

barambani opened the pull request:
Ref https://github.com/laserdisc-io/laserdisc/issues/76

it uses https://github.com/scala/scala/pull/2848 to reduce the allocation at extraction. Below some benchmarks

```
Benchmark                      Mode  Cnt        Score       Error  Units
ProtocolBench.decode (master) thrpt   20  2316529.598 ± 48239.316  ops/s
ProtocolBench.decode          thrpt   20  2228967.653 ± 40604.401  ops/s ~ -3.78%
ProtocolBench.decode (opt)    thrpt   20  2277232.571 ± 43841.897  ops/s ~ -1.70%
ProtocolBench.decode (opt 2)  thrpt   20  2350164.765 ± 41379.704  ops/s ~ +1.45%

ProtocolBench.encode (master) thrpt   20  3111032.093 ± 22264.895  ops/s
ProtocolBench.encode          thrpt   20  3081456.960 ± 60879.817  ops/s ~ -0.95%
ProtocolBench.encode (opt)    thrpt   20  3129469.697 ±  7551.217  ops/s ~ +0.59%
ProtocolBench.encode (opt 2)  thrpt   20  3075279.635 ±  9914.765  ops/s ~ -1.14%
```

barambani commented:
@sirocchj this is still WIP, I have to replace all the `"TODO: FILIPPO"` with the correct messages. Anyway you can have a first look.

sirocchj commented:
Looks great!! One thing: do we need the `EitherTestSyntax` when ScalaTest provides `EitherValues` (see http://www.scalatest.org/user_guide/other_goodies)

barambani commented:
You are absolutely right, it is provided but honestly it doesn't look very maintained at least wit our version of scalatests. What I mean is that it works on an implicit conversion of the `RightProjection` 
```scala
/**
   * Implicit conversion that adds a <code>value</code> method to <code>RightProjection</code>.
   *
   * @param either the <code>RightProjection</code> on which to add the <code>value</code> method
   */
  implicit def convertRightProjectionToValuable[L, R](rightProj: Either.RightProjection[L, R])(implicit pos: source.Position): RightValuable[L, R] = new RightValuable(rightProj, pos)
```
that unfortunately has been deprecated with the 2.13 as `Either` is right biased.
```scala
/** Projects this `Either` as a `Right`.
   *
   *  Because `Either` is right-biased, this method is not normally needed.
   */
  @deprecated("Either is now right-biased, use methods directly on Either", "2.13.0")
  def right = Either.RightProjection(this)
```
Honestly it's a lot of troubles only because the library is not keeping up. If you agree I would leave in place the syntax at least until we move to the `3.1.0` version (that by the way is blocked by this https://github.com/scalatest/scalatestplus-scalacheck/issues/17 for us). After that I will try again to use the provided `EitherValues` and we might even consider using a more modern testing lib.

sirocchj commented:
Yes please, let’s create a ticket to move to some testing framework a bit more modern

So this is almost good to go, you just need to come up with some nice error messages for those TODOs

barambani commented:
I went through another round of benchmarks and are as follows
```
Benchmark                        Mode  Cnt        Score       Error  Units
ProtocolBench.decode (master)   thrpt   20  2316529.598 ± 48239.316  ops/s
ProtocolBenchStr.decode (final) thrpt   20  2419916.666 ± 40979.383  ops/s ~ +4.46%

Benchmark                        Mode  Cnt        Score       Error  Units
ProtocolBench.encode (master)   thrpt   20  3111032.093 ± 22264.895  ops/s
ProtocolBenchStr.encode (final) thrpt   20  3149142.431 ± 61532.668  ops/s ~ +1.22%

Benchmark                                                Mode  Cnt        Score       Error  Units
ProtocolBenchArr.decodeArrOfArrOfBulk          (master) thrpt   20  17113.994 ±     292.633  ops/s
ProtocolBenchArr.decodeArrOfArrOfBulk          (final)  thrpt   20  16521.704 ±      67.897  ops/s

Benchmark                                                Mode  Cnt        Score       Error  Units
ProtocolBenchArr.decodeArrOfArrOfBulkWithNull  (master) thrpt   20  19080.265 ±     302.932  ops/s
ProtocolBenchArr.decodeArrOfArrOfBulkWithNull  (final)  thrpt   20  17767.420 ±     330.776  ops/s

Benchmark                                                Mode  Cnt        Score       Error  Units
ProtocolBenchArr.decodeArrOfBulk               (master) thrpt   20 883784.126 ±    4200.796  ops/s
ProtocolBenchArr.decodeArrOfBulk               (final)  thrpt   20 931495.585 ±   39441.643  ops/s

Benchmark                                                Mode  Cnt        Score       Error  Units
ProtocolBenchArr.decodeArrOfBulkWithNull       (master) thrpt   20 735011.180 ±   12753.885  ops/s
ProtocolBenchArr.decodeArrOfBulkWithNull       (final)  thrpt   20 947207.362 ±    3094.258  ops/s

Benchmark                                                Mode  Cnt        Score       Error  Units
ProtocolBenchArr.decodeArrOfPairs              (master) thrpt   20 164485.108 ±    2874.617  ops/s
ProtocolBenchArr.decodeArrOfPairs              (final)  thrpt   20 167526.315 ±    2501.701  ops/s
```
and the less relevant of them are
```
Benchmark                                                       Mode  Cnt           Score       Error  Units
RESPBench.baseline_utf8_decode                        (master) thrpt   20    612774.443 ±    9165.731  ops/s
RESPBench.baseline_utf8_decode                        (final)  thrpt   20    632846.968 ±   12935.305  ops/s
RESPBench.baseline_utf8_decodeI                       (master) thrpt   20    634374.825 ±   10221.663  ops/s
RESPBench.baseline_utf8_decodeI                       (final)  thrpt   20    633971.334 ±    8587.504  ops/s
RESPBench.baseline_utf8_encode                        (master) thrpt   20    250276.137 ±    1997.477  ops/s
RESPBench.baseline_utf8_encode                        (final)  thrpt   20    249398.619 ±    4031.292  ops/s
RESPBench.baseline_utf8_encodeI                       (master) thrpt   20    250839.784 ±     881.207  ops/s
RESPBench.baseline_utf8_encodeI                       (final)  thrpt   20    249652.080 ±    4084.043  ops/s

Benchmark                                                       Mode  Cnt           Score       Error  Units
RESPBench.bulk_decode                                 (master) thrpt   20    665909.359 ±   11351.152  ops/s
RESPBench.bulk_decode                                 (final)  thrpt   20    645283.843 ±    4875.437  ops/s
RESPBench.bulk_encode                                 (master) thrpt   20    792245.811 ±   12897.736  ops/s
RESPBench.bulk_encode                                 (final)  thrpt   20    788125.466 ±   12945.002  ops/s
RESPBench.err_decode                                  (master) thrpt   20   2314980.935 ±   86191.074  ops/s
RESPBench.err_decode                                  (final)  thrpt   20   2364638.290 ±   46067.282  ops/s
RESPBench.err_encode                                  (master) thrpt   20  11788318.086 ±   71770.300  ops/s
RESPBench.err_encode                                  (final)  thrpt   20  11673990.996 ±  216498.485  ops/s
RESPBench.num_decode                                  (master) thrpt   20   4042644.275 ±   72097.836  ops/s
RESPBench.num_decode                                  (final)  thrpt   20   4257526.722 ±   11665.013  ops/s
RESPBench.num_encode                                  (master) thrpt   20   9887938.304 ±  179527.879  ops/s
RESPBench.num_encode                                  (final)  thrpt   20   8552159.482 ±  293199.284  ops/s
RESPBench.str_decode                                  (master) thrpt   20   4145673.340 ±   12320.841  ops/s
RESPBench.str_decode                                  (final)  thrpt   20   5074156.875 ±   88346.653  ops/s
RESPBench.str_encode                                  (master) thrpt   20  12401657.688 ±  318416.867  ops/s
RESPBench.str_encode                                  (final)  thrpt   20  14524682.082 ±   87670.935  ops/s

Benchmark                                                       Mode  Cnt           Score       Error  Units
RESPParamWriteBench.write                             (master) thrpt   20   5031044.108 ±   82239.563  ops/s
RESPParamWriteBench.write                             (final)  thrpt   20   5168929.378 ±   13188.675  ops/s

Benchmark                                                       Mode  Cnt           Score       Error  Units
UTF8EncodingBench.decode_long_string_baseline_charset (master) thrpt   20    744917.563 ±    1751.992  ops/s
UTF8EncodingBench.decode_long_string_baseline_charset (final)  thrpt   20    702220.930 ±   13978.610  ops/s
UTF8EncodingBench.decode_long_string_baseline_string  (master) thrpt   20   1168714.375 ±    5240.169  ops/s
UTF8EncodingBench.decode_long_string_baseline_string  (final)  thrpt   20   1161153.017 ±   15950.488  ops/s
UTF8EncodingBench.decode_long_string_default_utf8     (master) thrpt   20    634016.948 ±    5343.047  ops/s
UTF8EncodingBench.decode_long_string_default_utf8     (final)  thrpt   20    635621.438 ±    1258.285  ops/s
UTF8EncodingBench.decode_long_string_lenient_utf8     (master) thrpt   20    967842.857 ±    3527.130  ops/s
UTF8EncodingBench.decode_long_string_lenient_utf8     (final)  thrpt   20    951877.513 ±   10283.922  ops/s
UTF8EncodingBench.decode_ok_baseline_charset          (master) thrpt   20  21303572.613 ±  356546.009  ops/s
UTF8EncodingBench.decode_ok_baseline_charset          (final)  thrpt   20  21417502.970 ±   85304.851  ops/s
UTF8EncodingBench.decode_ok_baseline_string           (master) thrpt   20  33258850.988 ± 1341843.920  ops/s
UTF8EncodingBench.decode_ok_baseline_string           (final)  thrpt   20  34649203.899 ±  139557.682  ops/s
UTF8EncodingBench.decode_ok_default_utf8              (master) thrpt   20  14471231.391 ±  106720.860  ops/s
UTF8EncodingBench.decode_ok_default_utf8              (final)  thrpt   20  14532025.447 ±  202423.602  ops/s
UTF8EncodingBench.decode_ok_lenient_utf8              (master) thrpt   20  18823210.673 ±  110296.949  ops/s
UTF8EncodingBench.decode_ok_lenient_utf8              (final)  thrpt   20  18771832.511 ±  223263.747  ops/s
UTF8EncodingBench.encode_long_string_baseline_charset (master) thrpt   20    259851.277 ±    3244.473  ops/s
UTF8EncodingBench.encode_long_string_baseline_charset (final)  thrpt   20    250430.579 ±    7839.620  ops/s
UTF8EncodingBench.encode_long_string_baseline_string  (master) thrpt   20    892231.624 ±    1018.138  ops/s
UTF8EncodingBench.encode_long_string_baseline_string  (final)  thrpt   20    886471.362 ±    6976.948  ops/s
UTF8EncodingBench.encode_long_string_default_utf8     (master) thrpt   20    249766.842 ±     429.907  ops/s
UTF8EncodingBench.encode_long_string_default_utf8     (final)  thrpt   20    250546.428 ±     774.390  ops/s
UTF8EncodingBench.encode_long_string_lenient_utf8     (master) thrpt   20    880794.583 ±    7419.475  ops/s
UTF8EncodingBench.encode_long_string_lenient_utf8     (final)  thrpt   20    882278.491 ±    4438.629  ops/s
UTF8EncodingBench.encode_ok_baseline_charset          (master) thrpt   20  26433620.529 ±  200314.816  ops/s
UTF8EncodingBench.encode_ok_baseline_charset          (final)  thrpt   20  26307496.442 ±  324216.757  ops/s
UTF8EncodingBench.encode_ok_baseline_string           (master) thrpt   20  29123295.091 ±  529493.162  ops/s
UTF8EncodingBench.encode_ok_baseline_string           (final)  thrpt   20  28040798.628 ±  139428.621  ops/s
UTF8EncodingBench.encode_ok_default_utf8              (master) thrpt   20  15325889.726 ±  226974.008  ops/s
UTF8EncodingBench.encode_ok_default_utf8              (final)  thrpt   20  14218022.220 ±  528313.536  ops/s
UTF8EncodingBench.encode_ok_lenient_utf8              (master) thrpt   20  21838735.918 ±   92462.348  ops/s
UTF8EncodingBench.encode_ok_lenient_utf8              (final)  thrpt   20  21804256.780 ±  226971.786  ops/s
```

sirocchj commented on core/src/test/scala/laserdisc/protocol/ClusterPSpec.scala:
```diff
@@ -50,7 +50,7 @@ final class ClusterPSpec extends BaseSpec with ClusterP {
     val rawNode = for {
       nid   <- nodeIdGen
-      h     <- Gen.oneOf(const(""), hostGen)
+      h     <- hostGen
```
I can't find now where this is documented (assuming it is) but I suspect that I added the possibility of hosts being empty strings is because I've seen it somewhere

barambani replied:
I will ad that case back.

barambani replied:
done

sirocchj commented on core/src/main/scala/laserdisc/protocol/KeyP.scala:
```diff
@@ -94,20 +96,21 @@ trait KeyBaseP {
+    case Str("set")    => Right(Some(KeyType.set))
+    case Str("zset")   => Right(Some(KeyType.zset))
+    case Str("hash")   => Right(Some(KeyType.hash))
+    case Str(_)        => Right(None)
```
isn't this changed semantics from before?
```suggestion
    case Str("none")   => Right(None)
    case Str(other)    => Left(Err(s"Unexpected string for Str ==> Option[KeyType]. Was $other"))
```

sirocchj commented on core/src/main/scala/laserdisc/protocol/ClusterP.scala:
```diff
@@ -32,62 +36,69 @@ object ClusterP {
+          case ("handshake", Right(fgs)) => Right(Flag.handshake :: fgs)
+          case ("noaddr", Right(fgs))    => Right(Flag.noaddress :: fgs)
+          case (other, Right(_))         => Left(Err(s"Unexpected flag encoding. Was $other"))
+          case (_, Left(e))              => Left(e)
```
```suggestion
          case (_, left)                 => left
```
or it doesn't typecheck?

sirocchj commented on core/src/main/scala/laserdisc/protocol/ClusterP.scala:
```diff
@@ -32,62 +36,69 @@ object ClusterP {
+              .flatMap {
+                _.split(SPACE_CH).toSeq match {
+                  case NodeId(id) +: A(Right(a)) +: Fs(Right(fs)) +: MM(Right(mm)) +:
+                        ToInt(NonNegInt(ps)) +: ToInt(NonNegInt(pr)) +: ToInt(NonNegInt(ce)) +: L(Right(l)) +: Ss(Right(ss)) =>
```
mmm aren't you still throwing away all those `Left`s you built above? :)

sirocchj commented on core/src/main/scala/laserdisc/protocol/ServerP.scala:
```diff
@@ -118,34 +130,37 @@ object ServerP {
+            s.split(LF * 2)
+              .flatMap {
+                _.split(LF_CH).toSeq match {
+                  case ISR(Right(infoSection)) +: PR(Right(parameters)) => Some(infoSection -> parameters)
```
^^ again, you're ignoring the `Left`s, should we maybe consider the above `private val x: A ==> B` (i.e. the `ISR` and `PR`) as `PartialFunction[A, B]` in case we can't do better than ignoring?

barambani replied:
this is wrong. I will fix it.

barambani commented on core/src/main/scala/laserdisc/protocol/Read.scala:
```diff
@@ -15,220 +15,304 @@ implicit final val myRead: Read[${A}, ${B}] = new Read[${A}, ${B}] {
-    case Inl(_)                    => None
+  ): Read[A :+: CNil, Option[B]] = Read.instance {
+    case Inl(i @ R(Right(b))) if !cond(i) => Right(Some(b))
+    case Inl(_)                           => Right(None)
```
this case describes the success in the decoding. I will add it.

barambani replied:
done

barambani commented on core/src/main/scala/laserdisc/protocol/ClusterP.scala:
```diff
@@ -8,11 +8,15 @@ object ClusterP {
-    implicit final val infoRead: Bulk ==> Info = Read.instancePF {
-      case Bulk(s) => new Info(s.split(CRLF).collect { case KVPairRegex(k, v) => k -> v }.toMap)
+    implicit final val infoRead: Bulk ==> Info = Read.instance {
+      case Bulk(s) => Right(new Info(s.split(CRLF).collect { case KVPairRegex(k, v) => k -> v }.toMap))
```
This also needs to be checked. It smells it needs a fix.

sirocchj commented on core/src/main/scala/laserdisc/protocol/Read.scala:
```diff
@@ -15,220 +15,304 @@ implicit final val myRead: Read[${A}, ${B}] = new Read[${A}, ${B}] {
-  final def instance[A, B](f: A => Option[B]): Read[A, B] = new Read[A, B] {
-    override def read(a: A): Option[B] = f(a)
-  }
-  final def instancePF[A, B](pf: PartialFunction[A, B]): Read[A, B] = instance(pf.lift)
```
Oh I agree wholeheartedly! I was trying to cater for cases where we have one failure case, like in most read instances defined in `Read.scala` and some other cases spread here and there, eg `pongRead`. When we have multiple reasons why we can fail the read and need to report them then I agree one would want to use `.instance`

barambani replied:
see https://github.com/barambani/laserdisc/pull/1

barambani commented:
It should be it. I think I have addressed all the comments. The changes to add back `instancePF` are here https://github.com/barambani/laserdisc/pull/1 and they can be merged into this. Let me know what you think.
